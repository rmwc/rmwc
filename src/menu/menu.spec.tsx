import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import {
  MenuSurfaceAnchor,
  Menu,
  MenuSurface,
  MenuItem,
  SimpleMenu,
  SimpleMenuSurface
} from './';

describe('Menu', () => {
  it('renders', () => {
    const { asFragment } = render(
      <Menu open onClose={() => {}}>
        <MenuItem>Cookies</MenuItem>
        <MenuItem>Pizza</MenuItem>
        <MenuItem>Icecream</MenuItem>
      </Menu>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('can be fixed', async () => {
    render(
      <MenuSurfaceAnchor>
        <Menu open fixed>
          <MenuItem>Cookies</MenuItem>
          <MenuItem>Pizza</MenuItem>
          <MenuItem>Icecream</MenuItem>
        </Menu>
      </MenuSurfaceAnchor>
    );

    expect(screen.getByRole('menu').parentElement).toHaveClass(
      'mdc-menu-surface--fixed'
    );
  });

  it('does not trigger action when disabled', async () => {
    const onSelect = jest.fn();

    render(
      <MenuSurfaceAnchor>
        <Menu open={true} onSelect={onSelect} onClose={jest.fn()}>
          <MenuItem>Cookies</MenuItem>
          <MenuItem disabled={true}>Pizza</MenuItem>
          <MenuItem>Icecream</MenuItem>
        </Menu>
      </MenuSurfaceAnchor>
    );

    userEvent.click(screen.getByText('Cookies'));

    await waitFor(() => expect(onSelect).toHaveBeenCalledTimes(1));

    userEvent.click(screen.getByText('Pizza'));

    await waitFor(() => expect(onSelect).toHaveBeenCalledTimes(1));

    userEvent.click(screen.getByText('Icecream'));

    await waitFor(() => expect(onSelect).toHaveBeenCalledTimes(2));
  });

  it('dynamically updates aria-hidden based on whether or not the menu is open', () => {
    const { container, rerender } = render(<Menu open />);

    expect(container.firstChild).toHaveAttribute('aria-hidden', 'false');

    rerender(<Menu />);

    expect(container.firstChild).toHaveAttribute('aria-hidden', 'true');
  });

  it('MenuSurface renders', () => {
    const { asFragment } = render(
      <MenuSurfaceAnchor>
        <button>Test</button>

        <MenuSurface open onClose={() => {}}>
          <MenuItem>Cookies</MenuItem>
          <MenuItem>Pizza</MenuItem>
          <MenuItem>Icecream</MenuItem>
        </MenuSurface>
      </MenuSurfaceAnchor>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('SimpleMenu renders', async () => {
    let val = 0;

    const { asFragment } = render(
      <SimpleMenu handle={<button>Test</button>} open onClose={() => val++}>
        <MenuItem>Cookies</MenuItem>
        <MenuItem>Pizza</MenuItem>
        <MenuItem>Icecream</MenuItem>
      </SimpleMenu>
    );

    userEvent.click(screen.getByText('Test'));

    await waitFor(() => {
      expect(val).toBe(1);
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it('Menu Item can be nested', () => {
    let val = null;

    const { asFragment } = render(
      <SimpleMenu
        handle={<button>Test</button>}
        open
        onSelect={(evt) => (val = evt.detail.index)}
      >
        <MenuItem>
          <span>Cookies</span>
        </MenuItem>
        <MenuItem>
          <span>Pizza</span>
        </MenuItem>
        <MenuItem>
          <span>Icecream</span>
        </MenuItem>
      </SimpleMenu>
    );

    userEvent.click(screen.getByText('Test'));

    expect(asFragment()).toMatchSnapshot();
  });

  it('SimpleMenuSurface renders', async () => {
    let val = 0;

    const { asFragment } = render(
      <SimpleMenuSurface
        handle={<button onClick={() => {}}>Test</button>}
        open
        onClose={() => {
          val++;
        }}
      >
        <div>Hello</div>
      </SimpleMenuSurface>
    );

    userEvent.click(screen.getByText('Test'));

    await waitFor(() => expect(val).toBe(1));

    expect(asFragment()).toMatchSnapshot();
  });

  it('can have custom classnames', () => {
    [MenuSurfaceAnchor, Menu, MenuItem].forEach(
      (Component: React.ComponentType<any>) => {
        const { container } = render(
          <Component className={'my-custom-classname'} />
        );
        expect(container.firstChild).toHaveClass('my-custom-classname');
      }
    );
  });
});
