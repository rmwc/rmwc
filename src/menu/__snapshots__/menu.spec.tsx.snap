// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu Menu Item can be nested 1`] = `
<DocumentFragment>
  <div
    class="mdc-menu-surface--anchor"
  >
    <div
      aria-hidden="false"
      class="mdc-menu mdc-menu-surface mdc-menu-surface--animating-open"
    >
      <ul
        class="mdc-list mdc-menu__items mdc-list"
        role="menu"
      >
        <li
          class="mdc-list-item"
          role="menuitem"
          tabindex="0"
        >
          <div
            class="mdc-list-item__ripple "
          />
          <span>
            Cookies
          </span>
        </li>
        <li
          class="mdc-list-item"
          role="menuitem"
          tabindex="0"
        >
          <div
            class="mdc-list-item__ripple "
          />
          <span>
            Pizza
          </span>
        </li>
        <li
          class="mdc-list-item"
          role="menuitem"
          tabindex="0"
        >
          <div
            class="mdc-list-item__ripple "
          />
          <span>
            Icecream
          </span>
        </li>
      </ul>
    </div>
    <button>
      Test
    </button>
  </div>
</DocumentFragment>
`;

exports[`Menu MenuSurface renders 1`] = `
<DocumentFragment>
  <div
    class="mdc-menu-surface--anchor"
  >
    <button>
      Test
    </button>
    <div
      class="mdc-menu-surface mdc-menu-surface--animating-open"
    >
      <li
        class="mdc-list-item"
        role="menuitem"
        tabindex="0"
      >
        <div
          class="mdc-list-item__ripple "
        />
        Cookies
      </li>
      <li
        class="mdc-list-item"
        role="menuitem"
        tabindex="0"
      >
        <div
          class="mdc-list-item__ripple "
        />
        Pizza
      </li>
      <li
        class="mdc-list-item"
        role="menuitem"
        tabindex="0"
      >
        <div
          class="mdc-list-item__ripple "
        />
        Icecream
      </li>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Menu SimpleMenu renders 1`] = `
<DocumentFragment>
  <div
    class="mdc-menu-surface--anchor"
  >
    <div
      aria-hidden="false"
      class="mdc-menu mdc-menu-surface mdc-menu-surface--animating-open mdc-menu-surface--open mdc-menu-surface--is-open-below"
      style="transform-origin: left top; left: 0px; top: 0px; max-height: 736px;"
    >
      <ul
        class="mdc-list mdc-menu__items mdc-list"
        role="menu"
      >
        <li
          class="mdc-list-item"
          role="menuitem"
          tabindex="0"
        >
          <div
            class="mdc-list-item__ripple "
          />
          Cookies
        </li>
        <li
          class="mdc-list-item"
          role="menuitem"
          tabindex="0"
        >
          <div
            class="mdc-list-item__ripple "
          />
          Pizza
        </li>
        <li
          class="mdc-list-item"
          role="menuitem"
          tabindex="0"
        >
          <div
            class="mdc-list-item__ripple "
          />
          Icecream
        </li>
      </ul>
    </div>
    <button>
      Test
    </button>
  </div>
</DocumentFragment>
`;

exports[`Menu SimpleMenuSurface renders 1`] = `
<DocumentFragment>
  <div
    class="mdc-menu-surface--anchor"
  >
    <div
      class="mdc-menu-surface mdc-menu-surface--animating-open mdc-menu-surface--open mdc-menu-surface--is-open-below"
      style="transform-origin: left top; left: 0px; top: 0px; max-height: 736px;"
    >
      <div>
        Hello
      </div>
    </div>
    <button>
      Test
    </button>
  </div>
</DocumentFragment>
`;

exports[`Menu renders 1`] = `
<DocumentFragment>
  <div
    aria-hidden="false"
    class="mdc-menu mdc-menu-surface mdc-menu-surface--animating-open"
  >
    <ul
      class="mdc-list mdc-menu__items mdc-list"
      role="menu"
    >
      <li
        class="mdc-list-item"
        role="menuitem"
        tabindex="0"
      >
        <div
          class="mdc-list-item__ripple "
        />
        Cookies
      </li>
      <li
        class="mdc-list-item"
        role="menuitem"
        tabindex="0"
      >
        <div
          class="mdc-list-item__ripple "
        />
        Pizza
      </li>
      <li
        class="mdc-list-item"
        role="menuitem"
        tabindex="0"
      >
        <div
          class="mdc-list-item__ripple "
        />
        Icecream
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
